plugins/destination/clickhouse/client/migrate.go

```diff
@@ -12,67 +14,140 @@ import (
 	eg.SetLimit(maxConcurrentMigrate)
 
-	for _, table := range queries.NormalizedTables(tables) {
+	for _, table := range newSchema {
 		table := table
-		eg.Go(func() (err error) {
+		eg.Go(func() error {
 			c.logger.Info().Str("table", table.Name).Msg("Migrating table started")
-			defer func() {
-				c.logger.Err(err).Str("table", table.Name).Msg("Migrating table done")
-			}()
-
 			if len(table.Columns) == 0 {
 				c.logger.Warn().Str("table", table.Name).Msg("Table with no columns, skip")
 				return nil
 			}
 
-			curr := current.Get(table.Name)
-			if curr == nil {
+			current := currentSchema.Get(table.Name)
+			if current == nil {
 				return c.createTable(ctx, table)
 			}
 
-			return c.autoMigrateTable(ctx, curr, table)
+			err := c.autoMigrate(ctx, table, current)
+			if err != nil {
+				return err
+			}
+			c.logger.Err(err).Str("table", table.Name).Msg("Migrating table done")
+			return nil
 		})
 	}
 
 	return eg.Wait()
 }
 
+func (c *Client) nonAutoMigrableTables(tables schema.Tables, currentTables schema.Tables) ([]string, [][]schema.TableColumnChange) {
+	var result []string
+	var tableChanges [][]schema.TableColumnChange
+	for _, t := range tables {
+		current := currentTables.Get(t.Name)
+		if current == nil {
+			continue
+		}
+		changes := t.GetChanges(current)
+		if !c.canSafeMigrate(changes) {
+			result = append(result, t.Name)
+			tableChanges = append(tableChanges, changes)
+		}
+	}
+	return result, tableChanges
+}
+
+func (*Client) canSafeMigrate(changes []schema.TableColumnChange) bool {
+	for _, change := range changes {
+		if change.Type == schema.TableColumnChangeTypeAdd && change.Current.CreationOptions.NotNull {
+			return false
+		}
+
+		if change.Type == schema.TableColumnChangeTypeRemove && change.Previous.CreationOptions.NotNull {
+			return false
+		}
+
+		if change.Type == schema.TableColumnChangeTypeUpdate {
+			return false
+		}
+	}
+	return true
+}
+
 func (c *Client) createTable(ctx context.Context, table *schema.Table) (err error) {
 	c.logger.Debug().Str("table", table.Name).Msg("Table doesn't exist, creating")
 
 	return c.conn.Exec(ctx, queries.CreateTable(table))
 }
 
-func (c *Client) autoMigrateTable(ctx context.Context, current, table *schema.Table) (err error) {
-	tableName := table.Name
-	c.logger.Info().Str("table", tableName).Msg("Table exists, migrating")
+func (c *Client) dropTable(ctx context.Context, table *schema.Table) (err error) {
+	c.logger.Debug().Str("table", table.Name).Msg("Dropping table")
+
+	return c.conn.Exec(ctx, queries.DropTable(table))
+}
 
-	have, want := current.GetChangedColumns(table)
-	if len(have) > 0 && c.spec.MigrateMode != specs.MigrateModeForced {
-		return fmt.Errorf("table %s has different types for columns %v but schema wants %v , add `migrate_mode: forced` to the destination spec to drop the columns", table.Name, have, want)
+func needsTableDrop(change schema.TableColumnChange) bool {
+	// We can safely add a nullable column without dropping the table
```

Comment:
See comment on the `canSafeMigrate` to transform to switch

Reply:
Simplified in https://github.com/cloudquery/cloudquery/pull/8255/commits/6d1bdd9e45554de6f372aa0519e567350cbf5448